File: cpp/src/blocking_queue_demo.cpp
// Demo driver for BlockingQueue -- see include/blocking_queue.hpp for the
// class itself (extracted there so tests/blocking_queue_test.cpp can
// exercise it directly).

#include "blocking_queue.hpp"
#include <optional>
#include <thread>
#include <iostream>

int main() {
    constexpr int ITEMS = 10;
    BlockingQueue<std::optional<int>> queue;

    std::thread producer([&] {
        for (int i = 1; i <= ITEMS; ++i) {
            queue.push(i);
            std::cout << "produced " << i << "\n";
        }
        queue.push(std::nullopt); // poison pill
    });

    std::thread consumer([&] {
        while (true) {
            auto item = queue.pop();
            if (!item.has_value()) break;
            std::cout << "                 consumed " << *item << "\n";
        }
    });

    producer.join();
    consumer.join();
    return 0;
}

File: java/src/BlockingQueueDemo.java
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;

/** Classic producer/consumer using a bounded blocking queue.
 *  put() blocks when full, take() blocks when empty -- no manual
 *  wait/notify or locking required. Compare with cpp/blocking_queue_demo.cpp,
 *  where this has to be built by hand from a mutex + condition_variable. */
public class BlockingQueueDemo {

    private static final int CAPACITY = 4;
    private static final int ITEMS = 10;
    private static final Object POISON_PILL = new Object();

    public static void main(String[] args) throws InterruptedException {
        BlockingQueue<Object> queue = new ArrayBlockingQueue<>(CAPACITY);

        Thread producer = new Thread(() -> {
            try {
                for (int i = 1; i <= ITEMS; i++) {
                    queue.put(i);
                    System.out.println("produced " + i);
                }
                queue.put(POISON_PILL); // signal completion
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }, "producer");

        Thread consumer = new Thread(() -> {
            try {
                while (true) {
                    Object item = queue.take();
                    if (item == POISON_PILL) break;
                    System.out.println("                 consumed " + item);
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }, "consumer");

        producer.start();
        consumer.start();
        producer.join();
        consumer.join();
    }
}
